content/blog/2019-12-18-halide-autoscheduler/index.md

```diff
@@ -0,0 +1,359 @@
+the traversal order for computing the function, and, for called functions,
+the granularity at which the function is stored and computed.
+
+Consider the schedule above for the functions `bx`, `by`, and `in`.
+Converted into Halide code, the schedule looks like the following:
+
+```
+by.reorder(y, x);
+
+bx.store_root();
+bx.compute_at(by, y);
+bx.reorder(y, x);
+```
+
+
+## Evaluation
+
+We evaluate the performance of the autoscheduler over three benchmarks.
+We do this by comparing the performance of the autoscheduled run
+(OPT configuration) vs. the run with the default schedule (DEF configuration).
+We measure runtime and memory usage using `gprof`.
+
+For the experiments, we set the weights for execution features as follows:
+
+Group       | Weight 
+------------|---------------
+mem         | 0.1
+loads       | 0.5
+stores      | 0.5
+arith ops   | 1.0
+math ops    | 10.0
+
+We run beam search with a depth of 10 and beam width of 300.
+
+For all benchmarks, the output functions are realized across an extent of
+2048x2048.
+The results below are averaged across three runs.
+
+### Benchmark 1
+
+```
+g(x,y) = sqrt(cos(x) + sin(y));
+f(x,y) = g(x + 1,y) + g(x,y) + g(x + 1,y + 1) + g(x,y + 1);
+```
+
+Function     | Runtime (ms) | Peak heap usage (bytes)
+-------------|--------------|------------------------
+f (DEF)      | 87.72        | 0
+f (OPT)      | 13.48        | 0
+g (DEF)      | 0.00         | 0
+g (OPT)      | 172.70       | 32874
+
+
+### Benchmark 2
+
+```
+blur_x(x,y) = input(x - 1,y) + input(x,y) + input(x + 1,y) / 3;
+blur_y(x,y) = blur_x(x - 1,y) + blur_x(x,y) + blur_x(x + 1,y) / 3;
+```
+
+Function     | Runtime (ms) | Peak heap usage (bytes)
+-------------|--------------|------------------------
+blur_y (DEF) | 12.70        | 0
+blur_y (OPT) | 19.02        | 0
+blur_x (DEF) | 0.00         | 0
+blur_x (OPT) | 16.16        | 16400
+
+
+### Benchmark 3
+
+```
+f(x,y) = x + y;
+```
+
+
+Function     | Runtime (ms) | Peak heap usage (bytes)
+-------------|--------------|------------------------
+f (DEF)      | 11.85        | 0
+f (OPT)      | 12.18        | 0
+
+
+### Discussion
+
+Note that for the DEF configuration, only the output functions have runtimes
+associated with them since all called functions are inlined.
+
+The autoscheduler performs rather poorly relative to the default schedule.
+While it successfully makes space-runtime tradeoffs (e.g. `f` in Benchmark 1),
+allowing the computation of a function to run much faster by saving
+intermediate results, it runs more slowly and uses more memory than the
+default schedule across all benchmarks.
+
+We believe the poor performance of the autoscheduler has two main causes:
+
+* **Wrong feature weights.** The feature weights for the cost model
+  are chosen by fiat; if these were learned instead given a set of
+  training data, then more the weights can probably better capture the
+  execution profile of schedules.
```

Comment:
Not that you should do this now, obviously, but you could imagine evaluating how well your current cost model correlates with real running times by just executing lots of the configurations you explored…